[NFC][Scalarizer][TargetTransformInfo] Add `isVectorIntrinsicWithOverloadTypeAtArg` api

inbelic opened the pull request:
This changes allows target intrinsic to specify and overwrite overloaded types.

This change will let us add scalarization for `asdouble`:  #114847

inbelic committed:
[NFC][Scalarizer][TargetTransformInfo] Add `isVectorIntrinsicWithOverloadTypeAtArg`

This changes allows target intrinsic to specify overloaded types.

This change will let us add scalarization for `asdouble`:

farzonl commented on llvm/lib/Transforms/Scalar/Scalarizer.cpp:
```diff
@@ -727,7 +727,8 @@ bool ScalarizerVisitor::splitCall(CallInst &CI) {
   SmallVector<llvm::Type *, 3> Tys;
   // Add return type if intrinsic is overloaded on it.
-  if (isVectorIntrinsicWithOverloadTypeAtArg(ID, -1))
+  if (TTI->isVectorIntrinsicWithOverloadTypeAtArg(
```
Passing in a default is a weird code pattern. also if you don't pass a target looking at TTI seems wrong. 
just do `isVectorIntrinsicWithOverloadTypeAtArg... && TTI->isVectorIntrinsicWithOverloadTypeAtArg...`

inbelic replied:
I agree, the problem is that the default behaviour is to only return `true` for `OpScalarIdx == -1`. Which in the `asdouble` case we don't want to add the `double` return type, so `||` doesn't work and when we want to add any other different index `&&` won't work.

farzonl replied:
I see. Still don't like the default behavior though. How about instead Create a helper in this file.  and check `Intrinsic::isTargetIntrinsic(ID)` if it is true do `TTI->isVectorIntrinsicWithOverloadTypeAtArg`. if it is false do `isVectorIntrinsicWithOverloadTypeAtArg`.

That way TTI isn't check unless we know its a target intrinsic.

inbelic committed:
review comments

- change code pattern to avoid calling TTI if we don't have a target
specific intrinsics
  - change type from unsigned to int

llvm-beanz reviewed (commented):
This feels like a partial change which leaves us in an odd state.

Instead of cleaning up all the uses of the VectorUtils `isVectorIntrinsicWithOverloadTypeAtArg`, you only updated the uses in the Scalarizer, which in turn makes the Scalarizer's implementation a bit odd.

farzonl reviewed (changes requested):
~~My mistake I thought we updated all the uses.~~

inbelic commented:
Hmm, okay. Are you suggesting that we should replace all the uses or that we can't do such a change?

Having looked at the other uses of `isVectorIntrinsicWithOverloadTypeAtArg`, they don't appear to have access to any target transform info ( eg `SLPVectorizer/ReplaceWithVeclib`) so I don't think there is a way forward to replace all the uses.

farzonl commented:
> This feels like a partial change which leaves us in an odd state.
> 
> Instead of cleaning up all the uses of the VectorUtils `isVectorIntrinsicWithOverloadTypeAtArg`, you only updated the uses in the Scalarizer, which in turn makes the Scalarizer's implementation a bit odd.

@llvm-beanz  I think I misunderstood what you were asking. We aren't implementing the TTI for all of `VectorUtils`. We don't have tests for target intrinsics that are vectorizable. Maybe one day it makes sense to add them for SPIRV?  But for DirectX keeping these changes in the Scalairzer is our path forward. It looks like @inbelic  did update all the Scalairzer uses.

llvm-beanz commented:
> @llvm-beanz I think I misunderstood what you were asking. We aren't implementing the TTI for all of `VectorUtils`.
I was not asking you to do that. I understand this is just one method.

> We don't have tests for target intrinsics that are vectorizable. Maybe one day it makes sense to add them for SPIRV? But for DirectX keeping these changes in the Scalairzer is our path forward. It looks like @inbelic did update all the Scalairzer uses.

I'm also unsure how this relates to my comment.

This change partially moves `isVectorIntrinsicWithOverloadTypeAtArg` from VectorUtils to TTI. It doesn't completely move it, it just half moves it... That seems like a really odd place to leave the code. The SLPVectorizer or VPlanRecipes not having TTI is a solvable problem.

It seems to me that we should have one way to query `isVectorIntrinsicWithOverloadTypeAtArg` across all passes, not two.

(cc @bogner)

inbelic committed:
review comments:

- consolidate the usage of isVectorIntrinsicWithOverloadTypeAtArg into
one place of VectorUtils.cppp

inbelic commented on llvm/include/llvm/Analysis/VectorUtils.h:
```diff
@@ -152,7 +152,8 @@ bool isVectorIntrinsicWithScalarOpAtArg(Intrinsic::ID ID,
 /// the operand at index \p OpdIdx, or on the return type if \p OpdIdx is -1.
-bool isVectorIntrinsicWithOverloadTypeAtArg(Intrinsic::ID ID, int OpdIdx);
+bool isVectorIntrinsicWithOverloadTypeAtArg(
+    Intrinsic::ID ID, int OpdIdx, const TargetTransformInfo *TTI = nullptr);
```
Could also do with an `std::optional` pointer here, not sure on what is preferred.

llvm-beanz replied:
I don't think this should be optional. In any case where this argument matters it would be a bug to not pass it.

bogner commented:
We really should be updating the other users of the `isVectorIntrinsicWithOverloadTypeAtArg` API to be passing in the `TTI`, or at least explicitly opting out of it if it makes sense. I think we should not have a default `nullptr` for the TTI argument here, as that makes it easy to misuse.

As for updating the vectorizer uses:
- SLPVectorizer should be easy to do, since the pass already uses TTI elsewhere
- VPlanRecipes is probably also not too hard, as it looks like TTI is used in some places there as well
- ReplaceWithVeclib would need to add a dependency on TTI. It might be okay to have that one pass in a null TTI though, since a target specific vectorizable function probably isn't going to be replaced with a generic vector library function anyway.

farzonl commented:
> We really should be updating the other users of the `isVectorIntrinsicWithOverloadTypeAtArg` API to be passing in the `TTI`, or at least explicitly opting out of it if it makes sense. I think we should not have a default `nullptr` for the TTI argument here, as that makes it easy to misuse.
> 
> As for updating the vectorizer uses:
> 
> * SLPVectorizer should be easy to do, since the pass already uses TTI elsewhere
> * VPlanRecipes is probably also not too hard, as it looks like TTI is used in some places there as well
> * ReplaceWithVeclib would need to add a dependency on TTI. It might be okay to have that one pass in a null TTI though, since a target specific vectorizable function probably isn't going to be replaced with a generic vector library function anyway.

Does it make sense to make a few issues for this and let this pr go in as is? I started this pattern with some of the other isVectorIntrinsic functions so feels like the cleanup should cover all the functions.

llvm-beanz commented:
> Does it make sense to make a few issues for this and let this pr go in as is? I started this pattern with some of the other isVectorIntrinsic functions so feels like the cleanup should cover all the functions.

It seems clear to me that we should file issues to fix the other cases, but I don't see why we would merge this one, then just have to clean it up. This is an NFC change, having the NFC refactoring end up in the correct end state seems better to me.

inbelic committed:
remove default TTI arg

inbelic committed:
update ReplaceWithVeclib uses

- this pass is used to replace a vectorizable intrinsic function with a
generic veclib function
- we make the assumption that a target specific intrinisc is not going
to be replaced with a generic function call and so we don't need to
provide the TTI

inbelic committed:
update uses in SLPVectorizer

- use the already defined TTI as the argument

inbelic committed:
update VPTransformState to include TargetTransformInfo

- pass down the TTI from the LoopVectorizationPlanner into the
VPTransformState
- this will allow this information to all VPRecipe::execute functions
when they generate new IR

inbelic committed:
update VPlanRecipe uses

- use the updated VPTransformState to pass down the TTI

inbelic commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -110,15 +110,17 @@ static bool replaceWithCallToVeclib(const TargetLibraryInfo &TLI,
   SmallVector<Type *, 3> OloadTys;
-  if (!RetTy->isVoidTy() && isVectorIntrinsicWithOverloadTypeAtArg(IID, -1))
+  if (!RetTy->isVoidTy() &&
+      isVectorIntrinsicWithOverloadTypeAtArg(IID, -1, /*TTI=*/nullptr))
```
Regarding adding a dependency on the `TargetTransformInfo` analysis pass to retrieve TTI. There is currently no code path that would execute and actually use `TTI` in the `VectorUtils.cpp` function. While I don't expect the pass is expensive to do, it is currently completely unnecessary work regardless.  So I think it is okay to defer until there is an actual use case. Which as mentioned, seems unlikely.

pow2clk reviewed (approved):
I have a comment suggestion you might consider, but looks good otherwise.

inbelic committed:
update comment

inbelic committed:
clang format
